### PR TITLE
Feature/add google authentication

### PR DESCRIPTION
**Pull request by jeremyjousse**
Add Google Auth mechanism to be able to call getToken.
But while calling getToken, it seems that requestBody is not empty, not using googleAuthRequest.

`Bad google token response: FailureResponse (Request {requestPath = (BaseUrl {baseUrlScheme = Https, baseUrlHost = "oauth2.googleapis.com", baseUrlPort = 443, baseUrlPath = ""},"/token"), requestQueryString = fromList [], requestBody = Just ((),application/x-www-form-urlencoded), requestAccept = fromList [application/json;charset=utf-8,application/json], requestHeaders = fromList [], requestHttpVersion = HTTP/1.1, requestMethod = "POST"}) (Response {responseStatusCode = Status {statusCode = 400, statusMessage = "Bad Request"}, responseHeaders = fromList [("Content-Type","application/json; charset=utf-8"),("Vary","Origin"),("Vary","X-Origin"),("Vary","Referer"),("Content-Encoding","gzip"),("Date","Sat, 25 Jul 2020 07:34:35 GMT"),("Server","scaffolding on HTTPServer2"),("Cache-Control","private"),("X-XSS-Protection","0"),("X-Frame-Options","SAMEORIGIN"),("X-Content-Type-Options","nosniff"),("Alt-Svc","h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""),("Transfer-Encoding","chunked")], responseHttpVersion = HTTP/1.1, responseBody = "{\n  \"error\": \"unsupported_grant_type\",\n  \"error_description\": \"Invalid grant_type: \"\n}"})`

I can give you securely GOOGLE_AUTH_CLIENT_ID, GOOGLE_AUTH_CLIENT_SECRET and GOOGLE_AUTH_REFRESH_TOKEN in order to test.

### CODE REVIEW

File: api/src/GoogleAuth.hs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+data GoogleAuthRequest = GoogleAuthRequest {grantType :: String,
+                                  clientId :: String,
+                                  clientSecret :: String,
+                                  refreshToken :: String,
+                                  redirectUri :: String}
+
+instance ToForm GoogleAuthRequest where 
+    toForm googleAuthRequest =
+        [ ("grant_ype", toQueryParam (grantType googleAuthRequest))
```
Comment (Sir4ur0n) [inline api/src/GoogleAuth.hs:95]: I think there is a typo, shouldn't it be `grant_type` rather than `grant_ype`?

**Comment by Sir4ur0n** [comment]
> responseBody = "{\n \"error\": \"unsupported_grant_type\",\n \"error_description\": \"Invalid grant_type: \"\n}"

This has me think the problem is more related to grant type for the moment, which is consistent with the typo in the field name :wink: Please first fix this and test again (I can't) and let me know what's going on

**Comment by Sir4ur0n** [comment]
(You have a conflict on the readme, maybe you haven't pulled `develop` recently on your branch?)

File: api/src/GoogleAuth.hs (hunk 1 above)
Comment (jeremyjousse) [reply api/src/GoogleAuth.hs:95]: What a shame having so bear fingers 😭
After fixing grant_type it works 🤟
Thanks!

**Comment by jeremyjousse** [comment]
> (You have a conflict on the readme, maybe you haven't pulled `develop` recently on your branch?)

Should be fixed, I merged from develop.

File: api/src/Lib.hs [hunk 2]
```diff
@@ -8,71 +8,75 @@ import           Effects.Database.Types
...
 mkPostgresApp = runStderrLoggingT $ withPostgresqlPool connStr 10 $ \pool -> do
-    liftIO $ runSqlPersistMPool (runMigration migrateAll) pool
-    pure $ app pool
+  liftIO $ runSqlPersistMPool (runMigration migrateAll) pool
+  pure $ app pool
 
 run :: IO ()
-run =
-  Warp.run 3000 =<< mkPostgresApp
+run = Warp.run 3000 =<< mkPostgresApp
```
Comment (Sir4ur0n) [inline api/src/Lib.hs:82]: I propose we rewrite this piece to make it more consistent (even though more verbose) with the rest of the code base for now. You can replace with:
```haskell
run :: IO ()
run = do
  application <- mkPostgresApp
  Warp.run 3000 application
```

Comment (Sir4ur0n) [reply api/src/Lib.hs:82]: We can always rework later, when you get more familiar with the various syntaxes for "binding". For now, let's stick to `do` notation only for bindings

File: api/src/Lib.hs [hunk 3]
```diff
@@ -8,71 +8,75 @@ import           Effects.Database.Types
...
-        case result of
-            Left errorMessage -> throwError $ err404 {errBody = encodeUtf8 $ fromStrict $ errorMessage}
-            Right deleteMessage -> pure deleteMessage
+  brandDelete :: (Key Brand) -> Handler Text
+  brandDelete id = do
+    result <- flip liftSqlPersistMPool pool $ do
+      mBrand <- selectFirst [BrandId ==. id] []
+      case mBrand of
+        Nothing ->
+          pure $ Left $ "The Band " <> pack (show id) <> " is not present"
```
[the remark this answers is not in the document]
Comment (Sir4ur0n) [reply api/src/Lib.hs]: Still not fixed

[the remark this answers is not in the document]
Comment (jeremyjousse) [reply api/src/Lib.hs]: Github is not perfect 😭
The file is really updated in the [add-google-authentication branch](https://github.com/jeremyjousse/mailparser/blob/feature/add-google-authentication/api/src/Lib.hs).
It is not the first time I have this issue.

[the remark this answers is not in the document]
Comment (Sir4ur0n) [reply api/src/Lib.hs]: In the file you linked (https://github.com/jeremyjousse/mailparser/blob/feature/add-google-authentication/api/src/Lib.hs) I also see 4 occurrences of `band`

[the remark this answers is not in the document]
Comment (jeremyjousse) [reply api/src/Lib.hs]: 😶 it's hard to live with a dyslexia.

**Pull request closed without merging**